src/lib.rs

```diff
@@ -41,7 +41,7 @@
 #![deny(broken_intra_doc_links)]
 
 #[cfg(any(all(feature = "std", feature = "no-std"), not(any(feature = "std", feature = "no-std"))))]
-compile_error!("exactly one of std and no-std must be enabled");
+compile_error!("either the `std` or `no-std` feature must be enabled");
```

Comment:
This looks wrong, Rust features are additive so the correct way to do it is enabling `std` with feature which may be default (although I personally dislike default features on libraries).

Reply:
the `std` feature *is* currently in the default set for the crate.  this check notifies the devs if they forgot to enable either `std` or `no-std` after disabling the default (we need one of them to provide an allocator and a hashmap implementation) or mistakenly enabled both

Reply:
AFAICT, the only reason we cant currently allow them to be freely set at the same time (ie use hashbrown with std, core2 with std (which is just std re-exported), etc, etc) is that bitcoin_hashes doesn't support building both std and no-std, but I think we should fix that bug in bitcoin_hashes instead and then we don't ever compile-error here.

Reply:
Right, we'll never get away from requiring *at least* no-std or std, but I feel like we should try hard to make sure we support *both* no-std and std.

Reply:
Ah, my confusion was caused by the `no-std` name. I believe it should be named `alloc`, not `no-std`. It'd then not seem contradictory to enable it while disabling `std`.

Regarding enabling both, I agree it should be possible. A little complication I see is that `HashSet` is in public API. Therefore it's not easy to determine what should happen if both features are enabled because either way could cause unrelated crates breaking each-other.  Fortunately the only way it's in public API is functions consuming it. Thus whole issue can be sidestepped by instead accepting generic `&impl Set<Item=Txid>`:

```rust
trait Set {
    type Item;

    fn contains(&T) -> bool;
}

#[cfg(feature = "std")]
impl Set for std::collections::HashSet { ... }

#[cfg(feature = "hashbrown")]
impl Set for hashbrown::HashSet { ... }
```

A somewhat interesting property of this may be the ability to use `BTreeMap` too. There's a small compatibility issue though: the code that relied on inference will stop compiling and will have to add `::<HashSet<_>>` to e.g. `collect()` - for instance the example needs to be fixed.

An alternative, if MSRV will be higher for `no_std`, is to pick whichever Rust version introduced the `alloc` crate. Then refer to `HashSet` by path `alloc::collections::HashSet`. This path will be the same as `std` so no breakage stemming from different types will happen if both are enabled. If neither is enabled, just disable the functions taking `HashSet`.

Reply:
I don't think we're ever going to get to a point and you can compile without at least either alloc or std. Note also that alloc does not provide a HashMap or HashSet (as they require ability to fetch a random number from the host).

Indeed, enabling both features with a different HashMap is an issue, but I like your proposed solution, I think.

Reply:
> they require ability to fetch a random number

Ah, forgot about this detail. So it seems the trait approach is the only clean solution I can see now.

Reply:
Ah! There's no actual reason for `Hash*` objects to be in the public interface, much easier to just remove them :)

Reply:
@TheBlueMatt had an idea to remove the `HashSet` from the public API and let hashbrown override the `HashSet` implementation if enabled rather than conflicting.  The latest commit now allows `std` + `no-std` to compile and keeps the API stable between the two features.

Reply:
@devrandom that's not right thing to do. It's a compatibility hazard.

Say crate A depends on `rust-bitcoin` and activates `hashbrown`.
Crate B depends on `rust-bitcoin` (without `hashbrown`) and calls the function with `std` version of `HashSet`

Crate C depends on A and B and can not compile because types will no longer match.

Reply:
Good point.

I adjusted the last commit to use the std `HashSet` for the `std` gated API methods.  Let me know if there's anything else.